scarlet/observation.py

```diff
@@ -0,0 +1,233 @@
+        self.padding = padding
+
+        if weights is not None:
+            self._weights = np.array(weights)
+        else:
+            self._weights = 1
+
+    def match(self, scene):
+
+        # 1) determine shape of scene in obs, set mask
+
+        # 2) compute the interpolation kernel between scene and obs
+
+        # 3) compute obs.psf in the frame of scene, store in Fourier space
+        # A few notes on this procedure:
+        # a) This assumes that scene.psfs and self.psfs have the same spatial shape,
+        #    which will need to be modified for multi-resolution datasets
+        if self._psfs is not None:
+            ipad, ppad = interpolation.get_common_padding(self._images, self._psfs, padding=self.padding)
+            self.image_padding, self.psf_padding = ipad, ppad
+            _psfs = np.pad(self._psfs, ((0, 0), *self.psf_padding), 'constant')
+            _target = np.pad(scene._psfs, self.psf_padding, 'constant')
+
+            new_kernel_fft = []
+            # Deconvolve the target PSF
+            target_fft = np.fft.fft2(np.fft.ifftshift(_target))
+
+            for _psf in _psfs:
+                observed_fft = np.fft.fft2(np.fft.ifftshift(_psf))
+                # Create the matching kernel
+                kernel_fft = observed_fft / target_fft
+                # Take the inverse Fourier transform to normalize the result
+                # Trials without this operation are slow to converge, but in the future
+                # we may be able to come up with a method to normalize in the Fourier Transform
+                # and avoid this step.
+                kernel = np.fft.ifft2(kernel_fft)
+                kernel = np.fft.fftshift(np.real(kernel))
+                kernel /= kernel.sum()
+                # Store the Fourier transform of the matching kernel
+                new_kernel_fft.append(np.fft.fft2(np.fft.ifftshift(kernel)))
+            self.psfs_fft = np.array(new_kernel_fft)
+
+        # 4) divide obs.psf from scene.psf in Fourier space
+
+        # [ 5) compute sparse representation of interpolation * convolution ]
+
+    @property
+    def images(self):
+        return self._images
+
+    @property
+    def weights(self):
+        if self._weights is None:
+            return None
+        elif self._weights == 1:
+            return self._weights
+        return self._weights
+
+    def get_model(self, model):
+        """Resample and convolve a model to the observation frame
+
+        Parameters
+        ----------
+        model: array
+            The model in some other data frame.
+
+        Returns
+        -------
+        model: array
+            The convolved and resampled `model` in the observation frame.
+        """
+        if self.wcs is not None:
+            msg = "get_model is currently only supported when the observation frame matches the scene"
+            raise NotImplementedError(msg)
+
+        def _convolve_band(model, psf_fft):
```

Comment:
Is there a reason why this is not an ordinary class method?

Reply:
I suppose that it could be. It's never used outside of `get_model` so I didn't see the need to make it available outside the method, but I can change that.

Reply:
it's inside of a method that gets called multiple times, so the interpreter has to run every time. Plus, it's easier to inherit from it when the method is available at the class level

Reply:
good point